src/Modules/OrchardCore.Commerce/Services/ProductService.cs

```diff
@@ -37,11 +38,42 @@ public async Task<IEnumerable<ProductPart>> GetProductsAsync(IEnumerable<string>
 
         var contentItems = await _contentManager.GetAsync(contentItemIds);
 
-        // We need BuildDisplayAsync to fill part.Elements with the fields. We could extract the logic from
-        // BuildDisplayAsync, but we would have to copy almost everything from there.
+        // We need to replicate some things that BuildDisplayAsync does to fill part.Elements with the fields. We can't
+        // use BuildDisplayAsync directly because it requires a BuildDisplayContext.
         foreach (var contentItem in contentItems)
         {
-            await _contentDisplayHandler.BuildDisplayAsync(contentItem, context: null);
+            var contentItemsPartName = contentItem.ContentType;
+
+            var contentItemsPartFields = _contentDefinitionManager.GetTypeDefinition(contentItem.ContentType).Parts
+                .FirstOrDefault(contentPartDefinition =>
+                    contentPartDefinition.PartDefinition.Name == contentItemsPartName)
+                .PartDefinition
+                .Fields;
+
+            var part = contentItem.Get<ContentPart>(contentItemsPartName);
+
+            foreach (var field in contentItemsPartFields)
+            {
+                var fieldName = field.Name;
+
+                switch (field.FieldDefinition.Name)
+                {
+                    case nameof(BooleanProductAttributeField):
+                        part.Get<BooleanProductAttributeField>(fieldName);
+                        break;
+
+                    case nameof(NumericProductAttributeField):
+                        part.Get<NumericProductAttributeField>(fieldName);
+                        break;
+
+                    case nameof(TextProductAttributeField):
+                        part.Get<TextProductAttributeField>(fieldName);
+                        break;
+
+                    default:
+                        continue;
+                }
+            }
```

Comment:
What is this for? You don't do anything with the results.

Reply:
Yes, but this is what makes it work (fill the fields in the content items). I looked at the source code and in that, this is what fills the fields.

Reply:
I can add a comment about it, but this part is essential.

Reply:
Strange, but okay. Anyway, we will certainly add more product attribute types in the future and then it will break. It should be done for every type that derives from `ProductAttributeField`.

Reply:
The problem is that I can't really check that, since the types of the fields are only given in `string`s. In the original code it uses generics.

Original code:
```
contentPart.Get<TField>(partFieldDefinition.Name);
```
I tried to use get with `ProductAttributeField` but it wants to create a C# object out of it and `ProductAttributeField` is an abstract class:
```
Type is an interface or abstract class and cannot be instantiated. Path 'PriceVariantsProduct.Size'.
```
A solution to `Get` all the fields (not relying on the fact that there are three `ProductAttributeField`s now) could be this:
```
part.Get(Type.GetType("OrchardCore.Commerce.Fields." + field.FieldDefinition.Name), fieldName);
```
What do you think?

Reply:
Maybe I could also get `"OrchardCore.Commerce.Fields."` dynamically, but I think that doesn't change.